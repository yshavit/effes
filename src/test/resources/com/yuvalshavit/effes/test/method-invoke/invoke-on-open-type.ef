type Bool = True | False

type Pet ?
  isHappy (fed: Bool) -> Bool

type Dog:
  is Pet
  isHappy (fed: Bool | Void) -> True: return True -- Dogs are always happy! wag wag wag

type Cat:
  is Pet
  isHappy (fed: Bool) -> Bool: return fed         -- This is not true! My cats are often happy, even when not eating.

feedThenCheck (pet: Pet), (whenFed: Bool) -> Bool:
  return pet.isHappy whenFed

main:
  print feedThenCheck(Dog, True)
  print feedThenCheck(Dog, False)
  print feedThenCheck(Cat, True)
  print feedThenCheck(Cat, False)
