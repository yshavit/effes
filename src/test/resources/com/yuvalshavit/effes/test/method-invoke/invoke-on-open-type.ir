def feedThenCheck (pet: Cat | Dog) (whenFed: False | True) -> False | True:
  ReturnStatement
    CaseExpression
      case:
        VarExpression
          arg: pet
          pos: 0
      of Cat:
        MethodInvoke
          method: Cat::isHappy
          arg0:
            CastExpression
              castTo: Cat
              VarExpression
                arg: pet
                pos: 0
          arg1:
            VarExpression
              arg: whenFed
              pos: 1
      of Dog:
        MethodInvoke
          method: Dog::isHappy
          arg0:
            CastExpression
              castTo: Dog
              VarExpression
                arg: pet
                pos: 0
          arg1:
            VarExpression
              arg: whenFed
              pos: 1
def Cat::isHappy (fed: False | True) -> False | True:
  ReturnStatement
    VarExpression
      arg: fed
      pos: 1
def Dog::isHappy (fed: False | True | Void) -> True:
  ReturnStatement
    CtorInvoke
      type: True
