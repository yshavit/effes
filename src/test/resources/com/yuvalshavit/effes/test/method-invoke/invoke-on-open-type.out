True
True
True
False
>> null
